Add NVMEM support

pdgendt opened the pull request:
Introduce the NVMEM subsystem. [Inspired from linux](https://docs.kernel.org/driver-api/nvmem.html) this subsystem allows to read and write data to Non-Volatile memory, for example to store a MAC address or SoC/Device specific data.

It comes with support for EEPROM devices. The device API backend is determined at runtime using `DEVICE_API_IS`.

See https://github.com/zephyrproject-rtos/zephyr/pull/96598 for an example use case.

Derived from https://github.com/zephyrproject-rtos/zephyr/pull/74835

talih0 committed:
devicetree: Add nvmem-consumer

Adds property nvmem-cells for referencing a node that stores some
configuration information. A typical use case is the reading of MAC
address from an EEPROM device.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>
Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: nvmem: api: Add test suite for NVMEM API

Add basic testing to validate the API of the NVMEM subsystem.
Add testing with an EEPROM backend.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
dts: bindings: nvmem: Add fixed-layout

Add a devicetree binding to represent a fixed layout in an NVMEM provider.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
nvmem: Introduce new subsystem for Non-Volatile Memory layer

This commit adds the NVMEM subsystem with a basic implementation for use
with EEPROM devices.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

decsny reviewed (dismissed):
please review linux binding https://www.kernel.org/doc/Documentation/devicetree/bindings/nvmem/ , note also the layouts folder

pdgendt committed:
MAINTAINERS: Add area for NVMEM

Create a section for the newly maintained NVMEM area.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

henrikbrixandersen reviewed (changes requested):
I like it! Simple, yet extensible and a good level of abstraction. A few comments below:

pdgendt committed:
tests: lib: devicetree: api: Add NVMEM test cases

Add test cases tot the devicetree API testsuite for NVMEM macros.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

henrikbrixandersen commented on include/zephyr/nvmem.h:
```diff
@@ -0,0 +1,330 @@
+#define NVMEM_CELL_INST_GET_BY_IDX_OR(inst, idx, default_value)                                    \
+	NVMEM_CELL_GET_BY_IDX_OR(DT_DRV_INST(inst), idx, default_value)
+
+#if defined(CONFIG_NVMEM) || defined(__DOXYGEN__)
```
Personally, I'd prefer this to fail at the linker stage instead of at runtime.

henrikbrixandersen reviewed (commented):
Apart from the yet-to-be-written documentation, this looks great.

mbolivar reviewed (commented):
Binding looks like a good start, no need to boil the ocean and support everything out of the box IMO.

pdgendt committed:
doc: services: Add NVMEM subsystem

Introduce a page for the Non-Volatile Memory subsystem with a reference to
the doxygen API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
doc: releases: 4.3: Add NVMEM entry

Add an entry to the API and options section for the Non-Volatile Memory
subsystem.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

GseoC commented:
Hello Pietr, thank you for this proposing this feature. Some minor comments.
I will try to make something from it in the near future as well.

nordicjm commented on include/zephyr/nvmem.h:
```diff
@@ -0,0 +1,354 @@
+ *
+ * @kconfig_dep{CONFIG_NVMEM}
+ *
+ * @return 0 on success, negative errno code on failure.
```
some of these use `@return` some use `@retval`, I vaugley remember that `@retval` needs multiple lines one per value, unsure on `@return` or if one should be used over the other

nordicjm commented on include/zephyr/devicetree/nvmem.h:
```diff
@@ -0,0 +1,310 @@
+ *			compatible = "fixed-layout";
+ *			#address-cells = <1>;
+ *			#size-cells = <1>;
+ *			mac_address: mac_address@fa {
```
this file also

henrikbrixandersen commented:
> Can this be used with SoC Flash, external NOR Flash?

Current implementation in this PR only supports EEPROMs as NVMEM providers, but this can very easily be extended in the future.

pdgendt commented:
> Can this be used with SoC Flash, external NOR Flash?

I guess it can, especially for read-only cells. The tricky bit would probably be writing, as it could require additional erase logic.
Not in scope of this PR though. I want to do the follow-up PR where the network drivers can use NVMEM cells for MAC address logic.

nashif commented:
@pdgendt  @henrikbrixandersen 

nice stuff, i have seen this on arch wg agenda last week, are there some minutes somewhere with the outcome?

mbolivar reviewed (approved):
Binding looks like a good start, no need to boil the ocean and support everything out of the box IMO.

pdgendt commented:
> @pdgendt @henrikbrixandersen
> 
> nice stuff, i have seen this on arch wg agenda last week, are there some minutes somewhere with the outcome?

Not sure if notes were taken, I presented the idea and what it is trying to solve (the MAC address logic in network drivers for me), feedback was very positive overall, as far as I can remember.

henrikbrixandersen commented:
My apologies, just realised I never got around to cleanup my notes from the meeting and actually post them.
